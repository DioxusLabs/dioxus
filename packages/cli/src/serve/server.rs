use crate::dioxus_crate::DioxusCrate;
use crate::serve::Serve;
use crate::{Error, Result};
use axum::extract::{Request, State};
use axum::middleware::{self, Next};
use axum::{
    body::Body,
    extract::{
        ws::{Message, WebSocket},
        Extension, WebSocketUpgrade,
    },
    http::{
        header::{HeaderName, HeaderValue, CACHE_CONTROL, EXPIRES, PRAGMA},
        Method, Response, StatusCode,
    },
    response::IntoResponse,
    routing::{get, get_service},
    Router,
};
use axum_server::tls_rustls::RustlsConfig;
use dioxus_cli_config::{Platform, WebHttpsConfig};
use dioxus_hot_reload::{DevserverMsg, HotReloadMsg};
use futures_channel::mpsc::{UnboundedReceiver, UnboundedSender};
use futures_util::stream;
use futures_util::{stream::FuturesUnordered, StreamExt};
use hyper::header::ACCEPT;
use hyper::HeaderMap;
use serde::{Deserialize, Serialize};
use std::net::TcpListener;
use std::path::Path;
use std::sync::Arc;
use std::sync::RwLock;
use std::{
    convert::Infallible,
    fs, io,
    net::{IpAddr, SocketAddr},
    process::Command,
};
use tokio::task::JoinHandle;
use tower::ServiceBuilder;
use tower_http::{
    cors::{Any, CorsLayer},
    services::fs::{ServeDir, ServeFileSystemResponseBody},
    ServiceBuilderExt,
};

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", content = "data")]
enum Status {
    ClientInit {
        application_name: String,
        platform: String,
    },
    Building {
        progress: f64,
        build_message: String,
    },
    BuildError {
        error: String,
    },
    Ready,
}

#[derive(Debug, Clone)]
struct SharedStatus(Arc<RwLock<Status>>);

impl SharedStatus {
    fn new(status: Status) -> Self {
        Self(Arc::new(RwLock::new(status)))
    }

    fn set(&self, status: Status) {
        *self.0.write().unwrap() = status;
    }

    fn get(&self) -> Status {
        self.0.read().unwrap().clone()
    }
}

pub(crate) struct Server {
    pub hot_reload_sockets: Vec<WebSocket>,
    pub build_status_sockets: Vec<WebSocket>,
    pub ip: SocketAddr,
    pub new_hot_reload_sockets: UnboundedReceiver<WebSocket>,
    pub new_build_status_sockets: UnboundedReceiver<WebSocket>,
    _server_task: JoinHandle<Result<()>>,
    /// We proxy (not hot reloading) fullstack requests to this port
    pub fullstack_port: Option<u16>,

    build_status: SharedStatus,
    application_name: String,
    platform: String,
}

impl Server {
    pub fn start(serve: &Serve, cfg: &DioxusCrate) -> Self {
        let (hot_reload_sockets_tx, hot_reload_sockets_rx) = futures_channel::mpsc::unbounded();
        let (build_status_sockets_tx, build_status_sockets_rx) = futures_channel::mpsc::unbounded();

        let build_status = SharedStatus::new(Status::Building {
            progress: 0.0,
            build_message: "Starting the build...".to_string(),
        });

        let addr = serve.server_arguments.address.address();
        let start_browser = serve.server_arguments.open.unwrap_or_default();

        // If we're serving a fullstack app, we need to find a port to proxy to
        let fullstack_port = if matches!(
            serve.build_arguments.platform(),
            Platform::Fullstack | Platform::StaticGeneration
        ) {
            get_available_port(addr.ip())
        } else {
            None
        };

        let fullstack_address = fullstack_port.map(|port| SocketAddr::new(addr.ip(), port));

        let router = setup_router(
            serve,
            cfg,
            hot_reload_sockets_tx,
            build_status_sockets_tx,
            fullstack_address,
            build_status.clone(),
        )
        .unwrap();

        // Actually just start the server, cloning in a few bits of config
        let web_config = cfg.dioxus_config.web.https.clone();
        let base_path = cfg.dioxus_config.web.app.base_path.clone();
        let _server_task = tokio::spawn(async move {
            let web_config = web_config.clone();
            // HTTPS
            // Before console info so it can stop if mkcert isn't installed or fails
            // todo: this is the only async thing here - might be nice to
            let rustls: Option<RustlsConfig> = get_rustls(&web_config).await.unwrap();

            // Open the browser
            if start_browser {
                open_browser(base_path, addr, rustls.is_some());
            }

            // Start the server with or without rustls
            if let Some(rustls) = rustls {
                axum_server::bind_rustls(addr, rustls)
                    .serve(router.into_make_service())
                    .await?
            } else {
                // Create a TCP listener bound to the address
                axum::serve(
                    tokio::net::TcpListener::bind(&addr).await?,
                    router.into_make_service(),
                )
                .await?
            }

            Ok(())
        });

        Self {
            hot_reload_sockets: Default::default(),
            build_status_sockets: Default::default(),
            new_hot_reload_sockets: hot_reload_sockets_rx,
            new_build_status_sockets: build_status_sockets_rx,
            _server_task,
            ip: addr,
            fullstack_port,

            build_status,
            application_name: cfg.dioxus_config.application.name.clone(),
            platform: serve.build_arguments.platform().to_string(),
        }
    }

    async fn send_build_status(&mut self) {
        let mut i = 0;
        while i < self.build_status_sockets.len() {
            let socket = &mut self.build_status_sockets[i];
            if send_build_status_to(&self.build_status, socket)
                .await
                .is_err()
            {
                self.build_status_sockets.remove(i);
            } else {
                i += 1;
            }
        }
    }

    pub async fn start_build(&mut self) {
        self.build_status.set(Status::Building {
            progress: 0.0,
            build_message: "Starting the build...".to_string(),
        });
        self.send_build_status().await;
    }

    pub async fn update_build_status(&mut self, progress: f64, build_message: String) {
        if !matches!(self.build_status.get(), Status::Building { .. }) {
            return;
        }
        self.build_status.set(Status::Building {
            progress,
            build_message,
        });
        self.send_build_status().await;
    }

    pub async fn send_hotreload(&mut self, reload: HotReloadMsg) {
        let msg = DevserverMsg::HotReload(reload);
        let msg = serde_json::to_string(&msg).unwrap();

        // Send the changes to any connected clients
        let mut i = 0;
        while i < self.hot_reload_sockets.len() {
            let socket = &mut self.hot_reload_sockets[i];
            if socket.send(Message::Text(msg.clone())).await.is_err() {
                self.hot_reload_sockets.remove(i);
            } else {
                i += 1;
            }
        }
    }

    /// Wait for new clients to be connected and then save them
    pub async fn wait(&mut self) -> Option<Message> {
        let mut new_hot_reload_socket = self.new_hot_reload_sockets.next();
        let mut new_build_status_socket = self.new_build_status_sockets.next();
        let mut new_message = self
            .hot_reload_sockets
            .iter_mut()
            .enumerate()
            .map(|(idx, socket)| async move { (idx, socket.next().await) })
            .collect::<FuturesUnordered<_>>();

        tokio::select! {
            new_hot_reload_socket = &mut new_hot_reload_socket => {
                if let Some(new_socket) = new_hot_reload_socket {
                    drop(new_message);
                    self.hot_reload_sockets.push(new_socket);
                    return None;
                } else {
                    panic!("Could not receive a socket - the devtools could not boot - the port is likely already in use");
                }
            }
            new_build_status_socket = &mut new_build_status_socket => {
                if let Some(mut new_socket) = new_build_status_socket {
                    drop(new_message);

                    // Update the socket with project info and current build status
                    let project_info = SharedStatus::new(Status::ClientInit { application_name: self.application_name.clone(), platform: self.platform.clone() });
                    if send_build_status_to(&project_info, &mut new_socket).await.is_ok() {
                        _ = send_build_status_to(&self.build_status, &mut new_socket).await;
                        self.build_status_sockets.push(new_socket);
                    }
                    return None;
                } else {
                    panic!("Could not receive a socket - the devtools could not boot - the port is likely already in use");
                }
            }
            Some((idx, message)) = new_message.next() => {
                match message {
                    Some(Ok(message)) => return Some(message),
                    _ => {
                        drop(new_message);
                        _ = self.hot_reload_sockets.remove(idx);
                    }
                }
            }
        }

        None
    }

    pub async fn send_build_error(&mut self, error: Error) {
        let error = error.to_string();
        self.build_status.set(Status::BuildError {
            error: ansi_to_html::convert(&error).unwrap_or(error),
        });
        self.send_build_status().await;
    }

    pub async fn send_reload(&mut self) {
        self.build_status.set(Status::Ready);
        self.send_build_status().await;
        for socket in self.hot_reload_sockets.iter_mut() {
            _ = socket
                .send(Message::Text(
                    serde_json::to_string(&DevserverMsg::FullReload).unwrap(),
                ))
                .await;
        }
    }

    /// Send a shutdown message to all connected clients
    pub async fn send_shutdown(&mut self) {
        for socket in self.hot_reload_sockets.iter_mut() {
            _ = socket
                .send(Message::Text(
                    serde_json::to_string(&DevserverMsg::Shutdown).unwrap(),
                ))
                .await;
        }
    }

    pub async fn shutdown(&mut self) {
        self.send_shutdown().await;
        for socket in self.hot_reload_sockets.drain(..) {
            _ = socket.close().await;
        }
    }

    /// Get the address the fullstack server should run on if we're serving a fullstack app
    pub fn fullstack_address(&self) -> Option<SocketAddr> {
        self.fullstack_port
            .map(|port| SocketAddr::new(self.ip.ip(), port))
    }
}

/// Sets up and returns a router
///
/// Steps include:
/// - Setting up cors
/// - Setting up the proxy to the endpoint specified in the config
/// - Setting up the file serve service
/// - Setting up the websocket endpoint for devtools
fn setup_router(
    serve: &Serve,
    config: &DioxusCrate,
    hot_reload_sockets: UnboundedSender<WebSocket>,
    build_status_sockets: UnboundedSender<WebSocket>,
    fullstack_address: Option<SocketAddr>,
    build_status: SharedStatus,
) -> Result<Router> {
    let mut router = Router::new();
    let platform = serve.build_arguments.platform();

    // Setup proxy for the endpoint specified in the config
    for proxy_config in config.dioxus_config.web.proxy.iter() {
        router = super::proxy::add_proxy(router, proxy_config)?;
    }

    // Setup base path redirection
    if let Some(base_path) = config.dioxus_config.web.app.base_path.clone() {
        let base_path = format!("/{}", base_path.trim_matches('/'));
        router = Router::new()
            .nest(&base_path, router)
            .fallback(get(move || async move {
                format!("Outside of the base path: {}", base_path)
            }));
    }

    // server the dir if it's web, otherwise let the fullstack server itself handle it
    match platform {
        Platform::Web => {
            // Route file service to output the .wasm and assets if this is a web build
            router = router.nest_service("/", build_serve_dir(serve, config));
        }
        Platform::Fullstack | Platform::StaticGeneration => {
            // For fullstack and static generation, forward all requests to the server
            let address = fullstack_address.unwrap();

            router = router.nest_service("/",super::proxy::proxy_to(
                format!("http://{address}").parse().unwrap(),
                true,
                |error| {
                    Response::builder()
                        .status(StatusCode::INTERNAL_SERVER_ERROR)
                        .body(Body::from(format!(
                            "Backend connection failed. The backend is likely still starting up. Please try again in a few seconds. Error: {:#?}",
                            error
                        )))
                        .unwrap()
                },
            ));
        }
        _ => {}
    }

    // Setup middleware to intercept html requests if the build status is "Building"
    router = router.layer(middleware::from_fn_with_state(
        build_status,
        build_status_middleware,
    ));

    // Setup websocket endpoint - and pass in the extension layer immediately after
    router = router.nest(
        "/_dioxus",
        Router::new()
            .route(
                "/",
                get(
                    |ws: WebSocketUpgrade, ext: Extension<UnboundedSender<WebSocket>>| async move {
                        ws.on_upgrade(move |socket| async move { _ = ext.0.unbounded_send(socket) })
                    },
                ),
            )
            .layer(Extension(hot_reload_sockets))
            .route(
                "/build_status",
                get(
                    |ws: WebSocketUpgrade, ext: Extension<UnboundedSender<WebSocket>>| async move {
                        ws.on_upgrade(move |socket| async move { _ = ext.0.unbounded_send(socket) })
                    },
                ),
            )
            .layer(Extension(build_status_sockets)),
    );

    // Setup cors
    router = router.layer(
        CorsLayer::new()
            // allow `GET` and `POST` when accessing the resource
            .allow_methods([Method::GET, Method::POST])
            // allow requests from any origin
            .allow_origin(Any)
            .allow_headers(Any),
    );

    Ok(router)
}

fn build_serve_dir(serve: &Serve, cfg: &DioxusCrate) -> axum::routing::MethodRouter {
    static CORS_UNSAFE: (HeaderValue, HeaderValue) = (
        HeaderValue::from_static("unsafe-none"),
        HeaderValue::from_static("unsafe-none"),
    );

    static CORS_REQUIRE: (HeaderValue, HeaderValue) = (
        HeaderValue::from_static("require-corp"),
        HeaderValue::from_static("same-origin"),
    );

    let (coep, coop) = match serve.server_arguments.cross_origin_policy {
        true => CORS_REQUIRE.clone(),
        false => CORS_UNSAFE.clone(),
    };

    let out_dir = cfg.out_dir();
    let index_on_404 = cfg.dioxus_config.web.watcher.index_on_404;

    get_service(
        ServiceBuilder::new()
            .override_response_header(
                HeaderName::from_static("cross-origin-embedder-policy"),
                coep,
            )
            .override_response_header(HeaderName::from_static("cross-origin-opener-policy"), coop)
            .and_then({
                let out_dir = out_dir.clone();
                move |response| async move { Ok(no_cache(index_on_404, &out_dir, response)) }
            })
            .service(ServeDir::new(out_dir)),
    )
    .handle_error(|error: Infallible| async move {
        (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("Unhandled internal error: {}", error),
        )
    })
}

fn no_cache(
    index_on_404: bool,
    out_dir: &Path,
    response: Response<ServeFileSystemResponseBody>,
) -> Response<Body> {
    // By default we just decompose into the response
    let mut response = response.into_response();

    // If there's a 404 and we're supposed to index on 404, upgrade that failed request to the index.html
    // We might want to isnert a header here saying we *did* that but oh well
    if response.status() == StatusCode::NOT_FOUND && index_on_404 {
        let body = Body::from(std::fs::read_to_string(out_dir.join("index.html")).unwrap());

        response = Response::builder()
            .status(StatusCode::OK)
            .body(body)
            .unwrap();
    };

    insert_no_cache_headers(response.headers_mut());
    response
}

pub fn insert_no_cache_headers(headers: &mut HeaderMap) {
    headers.insert(CACHE_CONTROL, HeaderValue::from_static("no-cache"));
    headers.insert(PRAGMA, HeaderValue::from_static("no-cache"));
    headers.insert(EXPIRES, HeaderValue::from_static("0"));
}

/// Returns an enum of rustls config
pub async fn get_rustls(web_config: &WebHttpsConfig) -> Result<Option<RustlsConfig>> {
    if web_config.enabled != Some(true) {
        return Ok(None);
    }

    let (cert_path, key_path) = match web_config.mkcert {
        Some(true) => get_rustls_with_mkcert(web_config)?,
        _ => get_rustls_without_mkcert(web_config)?,
    };

    Ok(Some(
        RustlsConfig::from_pem_file(cert_path, key_path).await?,
    ))
}

pub fn get_rustls_with_mkcert(web_config: &WebHttpsConfig) -> Result<(String, String)> {
    const DEFAULT_KEY_PATH: &str = "ssl/key.pem";
    const DEFAULT_CERT_PATH: &str = "ssl/cert.pem";

    // Get paths to store certs, otherwise use ssl/item.pem
    let key_path = web_config
        .key_path
        .clone()
        .unwrap_or(DEFAULT_KEY_PATH.to_string());

    let cert_path = web_config
        .cert_path
        .clone()
        .unwrap_or(DEFAULT_CERT_PATH.to_string());

    // Create ssl directory if using defaults
    if key_path == DEFAULT_KEY_PATH && cert_path == DEFAULT_CERT_PATH {
        _ = fs::create_dir("ssl");
    }

    let cmd = Command::new("mkcert")
        .args([
            "-install",
            "-key-file",
            &key_path,
            "-cert-file",
            &cert_path,
            "localhost",
            "::1",
            "127.0.0.1",
        ])
        .spawn();

    match cmd {
        Err(e) => {
            match e.kind() {
                io::ErrorKind::NotFound => tracing::error!("mkcert is not installed. See https://github.com/FiloSottile/mkcert#installation for installation instructions."),
                e => tracing::error!("an error occurred while generating mkcert certificates: {}", e.to_string()),
            };
            return Err("failed to generate mkcert certificates".into());
        }
        Ok(mut cmd) => {
            cmd.wait()?;
        }
    }

    Ok((cert_path, key_path))
}

pub fn get_rustls_without_mkcert(web_config: &WebHttpsConfig) -> Result<(String, String)> {
    // get paths to cert & key
    if let (Some(key), Some(cert)) = (web_config.key_path.clone(), web_config.cert_path.clone()) {
        Ok((cert, key))
    } else {
        // missing cert or key
        Err("https is enabled but cert or key path is missing".into())
    }
}

/// Open the browser to the address
pub(crate) fn open_browser(base_path: Option<String>, address: SocketAddr, https: bool) {
    let protocol = if https { "https" } else { "http" };
    let base_path = match base_path.as_deref() {
        Some(base_path) => format!("/{}", base_path.trim_matches('/')),
        None => "".to_owned(),
    };
    _ = open::that(format!("{protocol}://{address}{base_path}"));
}

fn get_available_port(address: IpAddr) -> Option<u16> {
    TcpListener::bind((address, 0))
        .map(|listener| listener.local_addr().unwrap().port())
        .ok()
}

/// Middleware that intercepts html requests if the status is "Building" and returns a loading page instead
async fn build_status_middleware(
    state: State<SharedStatus>,
    request: Request,
    next: Next,
) -> axum::response::Response {
    // If the request is for html, and the status is "Building", return the loading page instead of the contents of the response
    let accepts = request.headers().get(ACCEPT);
    let accepts_html = accepts
        .and_then(|v| v.to_str().ok())
        .map(|v| v.contains("text/html"));

    if let Some(true) = accepts_html {
        let status = state.get();
        if status != Status::Ready {
            let html = include_str!("../../assets/loading.html");
            return axum::response::Response::builder()
                .status(StatusCode::OK)
                // Load the html loader then keep loading forever
                // We never close the stream so any headless testing framework (like playwright) will wait until the real build is done
                .body(Body::from_stream(
                    stream::once(async move { Ok::<_, std::convert::Infallible>(html) })
                        .chain(stream::pending()),
                ))
                .unwrap();
        }
    }

    next.run(request).await
}

async fn send_build_status_to(
    build_status: &SharedStatus,
    socket: &mut WebSocket,
) -> Result<(), axum::Error> {
    let msg = serde_json::to_string(&build_status.get()).unwrap();
    socket.send(Message::Text(msg)).await
}
