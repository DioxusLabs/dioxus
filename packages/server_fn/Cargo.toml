[package]
name = "server_fn"
authors = ["Greg Johnston", "Ben Wishovich"]
license = "MIT"
repository = "https://github.com/leptos-rs/leptos"
description = "RPC for any web framework."
readme = "../README.md"
version = "0.8.5"
rust-version.workspace = true
edition.workspace = true

[dependencies]
throw_error = { workspace = true }
server_fn_macro_default = { workspace = true }
# used for hashing paths in #[server] macro
const_format = { workspace = true, default-features = true }
const-str = { workspace = true, default-features = true }
rustversion = { workspace = true, default-features = true }
xxhash-rust = { features = [
  "const_xxh64",
], workspace = true, default-features = true }
# used across multiple features
serde = { features = ["derive"], workspace = true, default-features = true }
send_wrapper = { features = [
  "futures",
], optional = true, workspace = true, default-features = true }
thiserror = { workspace = true, default-features = true }

# registration system
inventory = { optional = true, workspace = true, default-features = true }
dashmap = { workspace = true, default-features = true }

## servers
# actix
actix-web = { optional = true, workspace = true, default-features = false }
actix-ws = { optional = true, workspace = true, default-features = true }

# axum
axum = { optional = true, default-features = false, features = [
  "multipart",
], workspace = true }
tower = { optional = true, workspace = true, default-features = true }
tower-layer = { optional = true, workspace = true, default-features = true }

## input encodings
serde_qs = { workspace = true, default-features = true }
multer = { optional = true, workspace = true, default-features = true }

## output encodings
# serde
serde_json = { workspace = true, default-features = true }
serde-lite = { features = [
  "derive",
], optional = true, workspace = true, default-features = true }
futures = { workspace = true, default-features = true }
http = { workspace = true, default-features = true }
ciborium = { optional = true, workspace = true, default-features = true }
postcard = { features = [
  "alloc",
], optional = true, workspace = true, default-features = true }
hyper = { optional = true, workspace = true, default-features = true }
bytes = { workspace = true, default-features = true }
http-body-util = { optional = true, workspace = true, default-features = true }
rkyv = { optional = true, workspace = true, default-features = true }
rmp-serde = { optional = true, workspace = true, default-features = true }
base64 = { workspace = true, default-features = true }

# client
gloo-net = { optional = true, workspace = true, default-features = true }
js-sys = { optional = true, workspace = true, default-features = true }
wasm-bindgen = { workspace = true, optional = true, default-features = true }
wasm-bindgen-futures = { optional = true, workspace = true, default-features = true }
wasm-streams = { optional = true, workspace = true, default-features = true }
web-sys = { optional = true, features = [
  "console",
  "ReadableStream",
  "ReadableStreamDefaultReader",
  "AbortController",
  "AbortSignal",
], workspace = true, default-features = true }

# reqwest client
reqwest = { default-features = false, optional = true, features = [
  "multipart",
  "stream",
], workspace = true }
tokio-tungstenite = { optional = true, workspace = true, default-features = true }
url = { workspace = true, default-features = true }
pin-project-lite = { workspace = true, default-features = true }
tokio = { features = [
  "rt",
], optional = true, workspace = true, default-features = true }

[build-dependencies]
rustc_version = { workspace = true, default-features = true }

[dev-dependencies]
trybuild = { workspace = true, default-features = true }

[features]
axum-no-default = [
  "ssr",
  "generic",
  "dep:axum",
  "dep:hyper",
  "dep:http-body-util",
  "dep:tower",
  "dep:tower-layer",
]
form-redirects = []
actix-no-default = ["ssr", "dep:actix-web", "dep:actix-ws", "dep:send_wrapper"]
actix = ["actix-web/default", "actix-no-default"]
axum = ["axum/default", "axum-no-default", "axum/ws", "dep:tokio"]
browser = [
  "dep:gloo-net",
  "dep:js-sys",
  "dep:send_wrapper",
  "dep:wasm-bindgen",
  "dep:web-sys",
  "dep:wasm-streams",
  "dep:wasm-bindgen-futures",
]
serde-lite = ["dep:serde-lite"]
multipart = ["browser", "dep:multer"]
cbor = ["dep:ciborium"]
rkyv = ["dep:rkyv"]
msgpack = ["dep:rmp-serde"]
postcard = ["dep:postcard"]
default-tls = ["reqwest?/default-tls"]
rustls = ["reqwest?/rustls-tls", "tokio-tungstenite?/rustls"]
reqwest = ["dep:reqwest", "dep:tokio-tungstenite", "dep:tokio"]
ssr = ["inventory"]
generic = []

[package.metadata.docs.rs]
all-features = true
rustdoc-args = ["--generate-link-to-definition"]

# disables some feature combos for testing in CI
[package.metadata.cargo-all-features]
denylist = [
  "rustls",
  "default-tls",
  "form-redirects",
  "gloo-net",
  "js-sys",
  "wasm-bindgen",
  "web-sys",
  "tower",
  "tower-layer",
  "send_wrapper",
  "ciborium",
  "hyper",
  "inventory",
  "rkyv",
]
skip_feature_sets = [
  [
    "actix",
    "axum",
  ],
  [
    "actix",
    "generic",
  ],
  [
    "browser",
    "actix",
  ],
  [
    "browser",
    "axum",
  ],
  [
    "browser",
    "reqwest",
  ],
  [
    "browser",
    "generic",
  ],
  [
    "default-tls",
    "rustls",
  ],
  [
    "browser",
    "ssr",
  ],
  [
    "axum-no-default",
    "actix",
  ],
  [
    "axum-no-default",
    "browser",
  ],
  [
    "axum-no-default",
    "generic",
  ],
  [
    "rkyv",
    "json",
  ],
  [
    "rkyv",
    "cbor",
  ],
  [
    "rkyv",
    "url",
  ],
  [
    "rkyv",
    "serde-lite",
  ],
  [
    "url",
    "json",
  ],
  [
    "url",
    "cbor",
  ],
  [
    "url",
    "serde-lite",
  ],
  [
    "postcard",
    "json",
  ],
  [
    "postcard",
    "cbor",
  ],
  [
    "postcard",
    "url",
  ],
  [
    "postcard",
    "serde-lite",
  ],
  [
    "postcard",
    "rkyv",
  ],
]
max_combination_size = 2

[lints.rust]
unexpected_cfgs = { level = "warn", check-cfg = [
  'cfg(leptos_debuginfo)',
  'cfg(rustc_nightly)',
] }
