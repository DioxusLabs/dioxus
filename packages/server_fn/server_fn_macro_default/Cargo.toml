[package]
name = "server_fn_macro_default"
authors = ["Greg Johnston"]
license = "MIT"
repository = "https://github.com/leptos-rs/leptos"
description = "The default implementation of the server_fn macro without a context"
version = "0.8.5"
edition.workspace = true

[lib]
proc-macro = true

[dependencies]
syn = { workspace = true, default-features = true }
server_fn_macro_dioxus = { workspace = true }

[features]
