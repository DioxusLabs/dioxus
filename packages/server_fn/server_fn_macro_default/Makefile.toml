extend = { path = "../../cargo-make/main.toml" }

[tasks.check-format]
env = { LEPTOS_PROJECT_DIRECTORY = "../../" }
