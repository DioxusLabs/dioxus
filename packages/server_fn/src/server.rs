use crate::{
    request::Req,
    response::{Res, TryRes},
};
use std::future::Future;

/// A server defines a pair of request/response types and the logic to spawn
/// an async task.
///
/// This trait is implemented for any server backend for server functions including
/// `axum` and `actix-web`. It should almost never be necessary to implement it
/// yourself, unless youâ€™re trying to use an alternative HTTP server.
pub trait Server<Error, InputStreamError = Error, OutputStreamError = Error> {
    /// The type of the HTTP request when received by the server function on the server side.
    type Request: Req<
            Error,
            InputStreamError,
            OutputStreamError,
            WebsocketResponse = Self::Response,
        > + Send
        + 'static;

    /// The type of the HTTP response returned by the server function on the server side.
    type Response: Res + TryRes<Error> + Send + 'static;

    /// Spawn an async task on the server.
    fn spawn(
        future: impl Future<Output = ()> + Send + 'static,
    ) -> Result<(), Error>;
}
