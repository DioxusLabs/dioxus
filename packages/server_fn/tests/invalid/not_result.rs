use server_fn::{
    codec::JsonEncoding,
    error::{FromServerFnError, ServerFnErrorErr},
};
use server_fn_macro_default::server;

#[derive(
    Debug, thiserror::Error, Clone, serde::Serialize, serde::Deserialize,
)]
pub enum CustomError {
    #[error("error a")]
    A,
    #[error("error b")]
    B,
}

impl FromServerFnError for CustomError {
    type Encoder = JsonEncoding;

    fn from_server_fn_error(_: ServerFnErrorErr) -> Self {
        Self::A
    }
}

#[server]
pub async fn full_alias_result() -> CustomError {
    CustomError::A
}

fn main() {}
