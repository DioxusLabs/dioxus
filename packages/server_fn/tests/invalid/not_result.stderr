error[E0277]: CustomError is not a `Result` or aliased `Result`. Server functions must return a `Result` or aliased `Result`.
  --> tests/invalid/not_result.rs:25:1
   |
25 | #[server]
   | ^^^^^^^^^ Must return a `Result` or aliased `Result`.
   |
   = help: the trait `ServerFnMustReturnResult` is not implemented for `CustomError`
   = note: If you are trying to return an alias of `Result`, you must also implement `FromServerFnError` for the error type.
   = help: the trait `ServerFnMustReturnResult` is implemented for `Result<T, E>`
   = note: this error originates in the attribute macro `server` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: CustomError is not a `Result` or aliased `Result`. Server functions must return a `Result` or aliased `Result`.
  --> tests/invalid/not_result.rs:25:1
   |
25 | #[server]
   | ^^^^^^^^^ Must return a `Result` or aliased `Result`.
   |
   = help: the trait `ServerFnMustReturnResult` is not implemented for `CustomError`
   = note: If you are trying to return an alias of `Result`, you must also implement `FromServerFnError` for the error type.
   = help: the trait `ServerFnMustReturnResult` is implemented for `Result<T, E>`

error[E0271]: expected `impl Future<Output = CustomError>` to be a future that resolves to `Result<_, _>`, but it resolves to `CustomError`
  --> tests/invalid/not_result.rs:25:1
   |
25 | #[server]
   | ^^^^^^^^^ expected `Result<_, _>`, found `CustomError`
   |
   = note: expected enum `Result<_, _>`
              found enum `CustomError`
note: required by a bound in `ServerFn::run_body::{anon_assoc#0}`
  --> src/lib.rs
   |
   |     ) -> impl Future<Output = Result<Self::Output, Self::Error>> + Send;
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `ServerFn::run_body::{anon_assoc#0}`

error[E0277]: CustomError is not a `Result` or aliased `Result`. Server functions must return a `Result` or aliased `Result`.
  --> tests/invalid/not_result.rs:25:1
   |
25 | #[server]
   | ^^^^^^^^^ Must return a `Result` or aliased `Result`.
   |
   = help: the trait `ServerFnMustReturnResult` is not implemented for `CustomError`
   = note: If you are trying to return an alias of `Result`, you must also implement `FromServerFnError` for the error type.
   = help: the trait `ServerFnMustReturnResult` is implemented for `Result<T, E>`

error[E0308]: mismatched types
  --> tests/invalid/not_result.rs:25:1
   |
25 | #[server]
   | ^^^^^^^^^ expected `CustomError`, found `Result<_, _>`
   |
   = note: expected enum `CustomError`
              found enum `Result<_, _>`
help: consider using `Result::expect` to unwrap the `Result<_, _>` value, panicking if the value is a `Result::Err`
   |
25 | #[server].expect("REASON")
   |          +++++++++++++++++
